packages/zmarkdown/server/controllers/manifest.js

```diff
@@ -0,0 +1,116 @@
+      }
+
+      return
+    }
+
+    if (!responseSent) {
+      res.json([vfile.contents, vfile.data, vfile.messages])
+      responseSent = true
+    }
+  }
+
+  // Increment endpoint RPM
+  ioInc()
+
+  let processorType = req.body.output_type
+  let applyTemplate = false
+
+  // Latex Document is mirrored to LaTeX and
+  // a template is applied to the response.
+  if (processorType === 'latex-document') {
+    processorType = 'latex'
+    applyTemplate = true
+  }
+
+  const rawContent    = req.body.manifest
+  const options       = req.body.opts || {}
+
+  let parsedExtracts  = 0
+  const extracts      = []
+
+  // Flatten content into extracts using recursion
+  function appendToExtracts (children, depth) {
+    depth++
+
+    children.forEach(child => {
+      if (child.title) extracts.push({text: `# ${child.title}`, depth: depth - 1})
+      if (child.introduction) extracts.push({text: child.introduction, depth})
+
+      if (child.text) extracts.push({text: child.text, depth})
+
+      // Process element's children
+      if (child.children) appendToExtracts(child.children, depth)
+
+      if (child.conclusion) extracts.push({text: child.conclusion, depth})
+    })
+  }
+
+  // Start recursion by top-level element
+  appendToExtracts([rawContent], options.heading_shift || 0)
+
+  // Process flattened extracts simultaneously
+  extracts.forEach((extract, order) => {
+    const {text, depth} = extract
+    const localOptions = Object.assign({}, options, {heading_shift: depth})
+
+    const processor = processorFactory(processorType, localOptions, true)
+
+    processor(text, (e, vfile) => {
+      // If an error was found, we can return immediately
+      // The other parts of the content will continue to be
+      // processed, and generate errors in Sentry.
+      if (e) sendResponse(e, vfile)
+
+      extracts[order].vfile = vfile
+
+      // When all extracts are done (without errors), send the response
+      if (++parsedExtracts === extracts.length && !responseSent) {
+        // The assembler will automatically wipe off unneeded piece of data
+        const assembledVfile = extracts.map(e => e.vfile).reduce(vfileAssembler)
+
+        if (assembledVfile.messages) {
+          // Remove null-error messages
+          for (let i = 0; i < assembledVfile.messages.length; i++) {
+            if (!assembledVfile.messages[i]) {
+              assembledVfile.messages.splice(i)
+            }
+          }
+        } else {
+          assembledVfile.messages = []
+        }
+
+        // Apply the template if necessary
+        if (applyTemplate) {
+          const opts = Object.assign(options, {latex: assembledVfile.contents})
+
+          template(opts, (e, doc) => {
+            if (e) return sendResponse(e)
+
+            // Replace the content, and discard metadata, which have no meaning in LaTeX
+            Object.assign(assembledVfile, {contents: doc, data: {}})
+          })
+        }
+
+        sendResponse(null, assembledVfile)
```

Comment:
You have to wait for `template`'s callback to be done before sending the response, otherwise it might send `assembledVfile` before `Object.assign` (L108) happens.

Reply:
Refactored using `Promise.all` and fixed this.